### ISSUE TITLE
All games are no longer booting after April 2 build (190cef9) on Vulkan/Linux with Proton/Wine

### ISSUE THREAD

**Issue by Thaurin** [xenia-canary/xenia-canary#158]
### Validation

- [X] I've read the [FAQ](https://github.com/xenia-project/xenia/wiki/FAQ).
- [X] The Xenia build used is from the master branch (not MLBS/AlexVS/Canary/pull requests, etc.)
- [X] This issue isn't for tech support (help with Xenia).
- [X] If this issue occurs in a specific game, I've done analysis to locate the faulty subsystem of the emulator and a potential reason in it.
- [X] I've checked if this issue hasn't already been reported.
- [X] My device meets the minimum requirements: https://github.com/xenia-project/xenia/wiki/Quickstart#system-requirements
- [X] (If building) I have read the building doc: https://github.com/xenia-project/xenia/blob/master/docs/building.md

### Describe what's going wrong

I don't know if you're accepting reports of Xenia running under Proton/Wine on Linux at all, but since Xenia Canary works really well with Lost Odyssey on earlier builds as of at least March 13 `canary_experimental@8678becda`, I thought that maybe there could be something useful here anyway. **If not, I apologize profoundly; please close this issue.**

Running Xenia Canary `canary_experimental@9992d20` (and earlier builds), and mainline master, I get to the black screen with "Loading", but 100% of the time are then presented with an crash error. This is on Steam Deck hardware with the experimental Vulkan renderer.

Now, this error seems to originate from Wine (`../src-wine/dlls/winevulkan/loader_thunks.c`), so I'm not sure if it belongs here. Nonetheless, maybe the difference with the March 13 build working well with the same Proton versions could tell you something, or anything at all.

### Describe what should happen

On build `canary_experimental@8678becda` from March 13, the game loads without major issues and good performance (pretty stable 30 FPS) with only minor graphical glitching (red artifacts on the right side of the screen, occasional "noise" on character textures).

This is on Steam Deck hardware running the experimental Vulkan renderer.

## On Proton 8.0-1 and Experimental

    Assertion failed!

    Program: Z:\home\deckGames\xenia_latest\xenia_canary.exe
    File: ../src-wine/dlls/winevulkan/loader_thunks.c
    Line: 2703

    Expression: !status

    Press OK to exist the program, or Cancel to start the Wine debugger.

## On Proton 7.0-6 and 6.3-8

    Unhandled Exception in Xenia
    <Nested Exception Encountered>
    Faulting thread name: GPU Commands (010000004)

### If applicable, upload a logfile and link it here

https://pastebin.com/8buXzDBj

**Comment by exp111** [comment]
This problem exists on all games i've tested on the steam deck.
The latest build that doesn't crash is `canary_expermental@e9d1e51` (from april 2nd) with every build later (starting from `canary_experimental@5e0c674`) crashing either directly while starting a game or after a loading screen/intro.

If you press "Cancel" on the wine assertion prompt, you also get the `GPU Commands` prompt, so it's not dependant on the proton version.

**Comment by Thaurin** [comment]
The next tag in line is `190cef9`, which is only two commits ahead of `e9d1e51`:

    22dd2e52e Temporarily disable prefetchw check, odd report from user who definitely has prefetchw, but it is not being reported as present
    356300d14 Do not check if we should show the prefetchw error message if we already have an avx error message. mtmsrd writes the EE bit only now.

This seems to only make changes to this prefetchw check. Can you confirm that `190cef9` is showing the errors in this issue and `e9d1e51` isn't? If so, I may need to get xenia-canary compiled on my machine, even though I'm not good with C++. Seems to be managable, if true.

    $ git diff 190cef987..e9d1e51c3 --stat
     src/xenia/base/main_init_win.cc       | 16 ++++++----------
     src/xenia/cpu/ppc/ppc_emit_control.cc | 10 ++--------
     2 files changed, 8 insertions(+), 18 deletions(-)

**Comment by exp111** [comment]
Whoops seems like i didnt see `190cef9`.

Tested it and `190cef9` also runs fine. Meaning that whatever happened between the tags (https://github.com/xenia-canary/xenia-canary/compare/190cef9...5e0c674) is probably at fault.

There are some gpu related commits, so probably one of those.

**Comment by Thaurin** [comment]
I successfully compiled the canary branch, but when I tried `e9d1e51`, I got some warnings as errors. I think I'll try and fix the compile issues and see if I can pinpoint the exact code that introduced this.

**Comment by exp111** [comment]
Tested it and it runs fine without https://github.com/xenia-canary/xenia-canary/commit/c238d8af550c457ac61a4ebef6888b3895802dbb and https://github.com/xenia-canary/xenia-canary/commit/baa2ff78d8f44883b0d70ab89fd9b4ee53a45e80.

**Comment by lolcatplays** [comment]
> Tested it and it runs fine without [c238d8a](https://github.com/xenia-canary/xenia-canary/commit/c238d8af550c457ac61a4ebef6888b3895802dbb) and [baa2ff7](https://github.com/xenia-canary/xenia-canary/commit/baa2ff78d8f44883b0d70ab89fd9b4ee53a45e80).

Any way you could share that build?

**Comment by exp111** [comment]
> Any way you could share that build?

You can just use https://github.com/xenia-canary/xenia-canary/releases/tag/190cef9 as there isn't a big difference.

But here's my build of current master without those two commits:
[xenia_canary.zip](https://github.com/xenia-canary/xenia-canary/files/11506362/xenia_canary.zip)
Idk if it's okay to upload, remove if not (as people can just self-compile).

**Comment by Thaurin** [comment]
Did you just revert those two commits? I think I tried and got conflicts, which I did not try to fix as I am not familiar with the code.

**Comment by exp111** [comment]
Hmm i didnt get any conflicts. Nothing else was done iirc

**Comment by mbaran5** [comment]
https://github.com/xenia-canary/xenia-canary/issues/163 same here, didn't realize it was a previous issue. Closed mine

**Comment by Thaurin** [comment]
Ideally, a Xenia Canary developer could take a look at it and see if it's an easy fix or not. However, I don't know if they have Steam Deck hardware, or if this issue also exists on Vulkan on other Linux distros (probably) or if Vulkan on Linux is currently even considered.

**Comment by exp111** [comment]
@Thaurin probably best to rename the issue to be more generic as we know that this affects all games on at least the steam deck

**Comment by Thaurin** [comment]
> @Thaurin probably best to rename the issue to be more generic as we know that this affects all games on at least the steam deck

I'll change the title.

**Comment by All-StarsPixel** [comment]
> Tested it and it runs fine without [c238d8a](https://github.com/xenia-canary/xenia-canary/commit/c238d8af550c457ac61a4ebef6888b3895802dbb) and [baa2ff7](https://github.com/xenia-canary/xenia-canary/commit/baa2ff78d8f44883b0d70ab89fd9b4ee53a45e80).

how did you just remove those? could you please give me a tutorial of sorts?

**Comment by drewbaumann** [comment]
> > Tested it and it runs fine without [c238d8a](https://github.com/xenia-canary/xenia-canary/commit/c238d8af550c457ac61a4ebef6888b3895802dbb) and [baa2ff7](https://github.com/xenia-canary/xenia-canary/commit/baa2ff78d8f44883b0d70ab89fd9b4ee53a45e80).
> 
> how did you just remove those? could you please give me a tutorial of sorts?
```
git checkout master
git revert c238d8a baa2ff7
```

Then you would build it. I didn't look into those commands myself. I used @exp111 's copy.

**Comment by jasonmbrown** [comment]
Can anyone confirm that this Only occurs on the steamdeck? (I'm not great at c but I have been poking around to see what the actual error is.) I would love to know what Line: 2703 of the loader_thunks.c file actual is, but it appears to be generated on compiling proton/wine. Im thinking the issue is an issue with proton and the steamdeck architecture. The thunks file is generated by this function https://github.com/ValveSoftware/wine/blob/proton_8.0/dlls/winevulkan/make_vulkan#L2658 
using the vulkan API registry https://raw.githubusercontent.com/KhronosGroup/Vulkan-Docs/main/xml/vk.xml Although I believe that one is a newer one compared to the one used with current proton versions.

Hopefully at some point a newer version of proton will fix the issue, but as I am lazy and don't want to bother attempting to generate the loader_thunks.c file (at least right now), I cant tell what the issue really is.

TL;DR;
I'm dumb, The issue is most likely a proton/wine Vulkan Api issue. The issue is most likely Steamdeck only (given the reports of this error seem to all be steamdeck related), meaning it is possibly a vulkan api binding that has not yet been implemented for the Steamdecks apu.

Final Note:
If anyone has the loader_thunks.c file and wants to send it to me (so I don't have to struggle setting up a c environment to compile) I might be able to get a bit more insight as to what is actually happening and why c238d8a and baa2ff7 break xenia.

Edit: After looking at the changes from c238d8a and baa2ff7 it really doesn't seem to make any sense as to why those break proton on the steamdeck...

**Comment by Thaurin** [comment]
> If anyone has the loader_thunks.c file and wants to send it to me (so I don't have to struggle setting up a c environment to compile) I might be able to get a bit more insight as to what is actually happening and why https://github.com/xenia-canary/xenia-canary/commit/c238d8af550c457ac61a4ebef6888b3895802dbb and https://github.com/xenia-canary/xenia-canary/commit/baa2ff78d8f44883b0d70ab89fd9b4ee53a45e80 break xenia.

Google led me to this commit in the [Wine repo for version 8.10 from 5 days ago](https://source.winehq.org/git/wine.git/blob/68a3b0077e64d1b5232ff75996b82766bcc64ced:/dlls/winevulkan/loader_thunks.c) where the file is located at `dlls/winevulkan/loader_thunks.c`. So maybe not generated?

Interestingly, it seems the error also happens for some people in [certain steam games, like Street Fighter 6](https://steamcommunity.com/app/1364780/discussions/0/3815158655062609525/), so likely not a Xenia issue but Vulkan, as you said.

Hope this `loader_thunks.c` can tell you something. It wasn't that hard to find.

> struggle setting up a c environment

Wasn't that much of a struggle, either, but I already had Visual Studio 2022, git, and so on installed.

> Edit: After looking at the changes from https://github.com/xenia-canary/xenia-canary/commit/c238d8af550c457ac61a4ebef6888b3895802dbb and https://github.com/xenia-canary/xenia-canary/commit/baa2ff78d8f44883b0d70ab89fd9b4ee53a45e80 it really doesn't seem to make any sense as to why those break proton on the steamdeck...

Indeed, it seems weird that these seem to have introduced it when Xenia worked well enough before and the issue is with Vulkan.

**Comment by jasonmbrown** [comment]
Thanks Thaurin, I'm not sure why I had trouble finding that file, but it was 2am so I was pretty tired. It looks like If I want to get any deeper I will need to learn what the value of the 'status' variable when the assertion is made (line 2711). Hopefully it wont be too difficult to figure out. 

Ok so the error that is returned by status is an NT Permissions error... so as of now I got no idea why or how to fix. I attempted to run a crashing 360 iso using a bunch of the proton compatibilities (individually) and also tried running it in protonGE but no luck whatsoever. 

It very well could be a kernel issue rather then an api issue, and I think thats something valve would need to fix.

**Comment by Thaurin** [comment]
I linked to Wine 8.10, but should've linked to 8.0.1, because that's the Proton version the original issue was written for. The failed assertion is on line 2703:

    status = UNIX_CALL(vkCreateGraphicsPipelines, &params);
    assert(!status);

Vulkan seems to have trouble creating graphics pipelines?

**Comment by jasonmbrown** [comment]
Ah I made a mistake with the line my bad, but the question remains, why the changes from c238d8a and baa2ff7 broke the function call. I would most likely need to debug on a crashing system or maybe get a memory dump (Basically just to see the difference of the &param values before and after c238d8a and baa2ff7.)

I haven't looked too in depth at the code, having a stack trace would help a bit. But without setting up a debugger the issue is beyond my capabilities. At best I can just speculate and poke through the code looking for an answer.

I'm also unsure if this should be considered an issue with wine or xenia at this point.

**Comment by ChalicerK** [comment]
I don't know if that helps you but something you could explore as well - at least for the steam deck.

I had an assertion failure with loader_thunks.c
what happened lately was the release of the new steam overlay.
https://steamcommunity.com/groups/SteamClientBeta/discussions/0/6861841362662276072/

By disabling the new steam overlay the game started again.

**Comment by superboo07** [comment]
still an issue, and disabling the steam overlay is not an option on steam deck. Please fix

**Comment by Towfieee** [comment]
under proton experiment component in steam select the bleeding edge release and not the "untested". This worked for me.

**Comment by Margen67** [comment]
Does this happen on master?

**Comment by MastaG** [comment]
So this Xenia can't be compiled to run on Linux natively?

**Comment by prince-rakeem** [comment]
I am having the same issue in Batocera 37 running on a x86_64 mini PC when attempting to run any game.

Xenia canary <nested exception encountered> Faulting thread name: GPU commands (01000004)

According to Batocera 37 change log this is `1319ff6` (Commits on May 2, 2023).

**Comment by Bacon777** [comment]
> > Any way you could share that build?
> 
> You can just use https://github.com/xenia-canary/xenia-canary/releases/tag/190cef9 as there isn't a big difference.
> 
> But here's my build of current master without those two commits: [xenia_canary.zip](https://github.com/xenia-canary/xenia-canary/files/11506362/xenia_canary.zip) Idk if it's okay to upload, remove if not (as people can just self-compile).

Any chance you have a more recent compiled version? Or are you still running this same one? Thank you

**Comment by purplebar0** [comment]
On AMD, the assertion failure issue only appears to happen on RADV, whereas on AMDVLK I could get RDR1 to get in-game as a test.

While it is possible to install AMDVLK and keep it as the default AMD driver, I recommend that instead of that, you disable AMDVLK's switching system as well (there are edge cases where the switcher is buggy). If you don't want to do any of these things, only install AMDVLK and roll with it.

---

In order to disable the switching system, add this to `/etc/environment`:

```
DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1=1
VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/radeon_icd.i686.json:/usr/share/vulkan/icd.d/radeon_icd.x86_64.json
```

Then, make sure that this environment variable is set for Xenia (running under Wine):

```
VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/amd_icd32.json:/usr/share/vulkan/icd.d/amd_icd64.json
```

---

If you don't want to disable the switcher, however, add this to `/etc/environment` _instead_:

```
AMD_VULKAN_ICD=RADV
```

Then, set this environment variable for Xenia:

```
AMD_VULKAN_ICD=amdvlk
```

**Comment by popsUlfr** [comment]
Here's a build with the two commits reverted based on latest HEAD dc29307

[xenia_canary.linuxfix.dc29307.zip](https://github.com/xenia-canary/xenia-canary/files/12385551/xenia_canary.linuxfix.dc29307.zip)

Still manages to work around the problem.

**Comment by Bacon777** [comment]
> On AMD, the assertion failure issue only appears to happen on RADV, whereas on AMDVLK I could get RDR1 to get in-game as a test.
> 
> While it is possible to install AMDVLK and keep it as the default AMD driver, I recommend that instead of that, you disable AMDVLK's switching system as well (there are edge cases where the switcher is buggy). If you don't want to do any of these things, only install AMDVLK and roll with it.
> ```

The only game I have been able to get to work is the 700 Goldeneye dump. Not sure if that helps in anyway. It boots up without any issues or tweaking. I will try to make those changes after playing with @popsUlfr's build.

> Here's a build with the two commits reverted based on latest HEAD [dc29307](https://github.com/xenia-canary/xenia-canary/commit/dc29307a559fe0bf35ab1157e6c4768a82cccf27)
> 
> [xenia_canary.linuxfix.dc29307.zip](https://github.com/xenia-canary/xenia-canary/files/12385551/xenia_canary.linuxfix.dc29307.zip)
> 
> Still manages to work around the problem.

Thank you, I will report back if the build is successful for my deck.

**Comment by popsUlfr** [comment]
> Thank you, I will report back if the build is successful for my deck.

No problem :)
I built this in a Win 11 virtual machine running on my Steam Deck (true story haha).
Can confirm Red Dead Redemption works on my Steam Deck with this build.

**Comment by SeongGino** [comment]
FWIW Xenia boots fine on my Arch/NVIDIA system with Wine-GE, and with the recent VKD3D 2.10 release the D3D12 backend now works (and muuuuch faster than Xenia's built-in Vulkan backend). Seems about on-par now with how it should perform on Windows.
Can't speak though for Deck or AMD unfortunately, unless others would like to confirm this?

**Comment by Thesola10** [comment]
@SeongGino absolutely, on Proton Experimental it's a blast, solid performance on Forza Motorsport 4 for instance

**Comment by mbaran5** [comment]
Are you still using canary or main line Xenia?

**Comment by SeongGino** [comment]
Canary; I hadn't tried Xenia from the master branch (and didn't see a need to, tbh)

**Comment by Thesola10** [comment]
@mbaran5 it also works on Stable, I use it for eg. joyRide Turbo which fails to go past Microsoft Studios screen on Canary

**Comment by andyktaylor** [comment]
Anyone with a Steamdeck, are you using the official build or the one popsUlfr posted?

What are your settings in the app? D3D12 or Vulkan? Any other tweaks needed from the default?

**Comment by coatlessali** [comment]
Is it related to #139?

> This fixes a bug where a stencil reference had the wrong type. Other driver implementations seem not to care, however, RADV does.

This is what fixed Xenia on RADV last time.

**Comment by langestefan** [comment]
> Anyone with a Steamdeck, are you using the official build or the one popsUlfr posted?
> 
> What are your settings in the app? D3D12 or Vulkan? Any other tweaks needed from the default?

The one that popsUlfr posted works great. No special settings, kept everything default. Using vulkan on steamdeck.

**Comment by Thaurin** [comment]
> The one that popsUlfr posted works great. No special settings, kept everything default. Using vulkan on steamdeck.

Try using DirectX 12 instead of Vulkan. Somewhere along the line, it got fixed either in Xenia or Mesa. It seems to be on par with DirectX 12 on Windows now.

**Comment by andyktaylor** [comment]
> > The one that popsUlfr posted works great. No special settings, kept everything default. Using vulkan on steamdeck.
> 
> Try using DirectX 12 instead of Vulkan. Somewhere along the line, it got fixed either in Xenia or Mesa. It seems to be on par with DirectX 12 on Windows now.

Whilst this is true, Gears of War 2 and 3 both freeze if DX12 so need Vulkan to work

**Comment by amarekosistem** [comment]
on batocera 38 with AMD 6700XT error :
Xenia canary Faulting thread name: GPU commands (01000004)
Xenia master its working, but some game like gears of war got transparent issue glitch

On intel mini pc i7 Inter Iris XE and linux Fedora with lutris with runner lutris-GE-Proton8-15-x85_64 :
Xenia Canary running well no error GPU commands
Xenia master not tested, because no point to test it as it already run with xenia canary


Both using Vulkan
Hopefully all problem can be sorted, thank you guys!!

**Comment by andyktaylor** [comment]
> Here's a build with the two commits reverted based on latest HEAD [dc29307](https://github.com/xenia-canary/xenia-canary/commit/dc29307a559fe0bf35ab1157e6c4768a82cccf27)
> 
> [xenia_canary.linuxfix.dc29307.zip](https://github.com/xenia-canary/xenia-canary/files/12385551/xenia_canary.linuxfix.dc29307.zip)
> 
> Still manages to work around the problem.

Any chance of a build using the latest version please?

**Comment by coatlessali** [comment]
Vulkan is working for me on AMD again with #244. On top of that, it seems to have fixed the weird black voids on Intel.

**Comment by midnightcatwalk** [comment]
Seeing a similar problem again after today’s Proton Experimental update on Steam Deck. The failed assertion this time is at `\windows\system32\steam.exe`, and the file in question is `steamclient_main.c` line 624. Expression: “!status”
